code/game/machinery/autolathe.dm

```diff
@@ -61,104 +60,126 @@
 	else
-		flick("autolathe_o", src)//plays metal insertion animation
-
-		use_power(min(1000, amount_inserted / 100))
-	updateUsrDialog()
+		data["designs"] = handle_designs(matching_designs, FALSE)
+	return data
+
+/obj/machinery/autolathe/proc/handle_designs(list/designs, categorycheck)
+	var/list/output = list()
+	for(var/v in designs)
+		var/datum/design/D = categorycheck ? SSresearch.techweb_design_by_id(v) : v
+		if(categorycheck)
+			if(!(selected_category in D.category))
+				continue
+		var/unbuildable = FALSE // we can't build the design currently
+		var/m10 = FALSE // 10x mult
+		var/m25 = FALSE // 25x mult
+		var/m50 = FALSE // 50x mult
+		var/m5 = FALSE // 5x mult
+		var/sheets = FALSE // sheets or no?
+		if(disabled || !can_build(D))
+			unbuildable = TRUE
+		var/max_multiplier = unbuildable ? 0 : 1
+		if(ispath(D.build_path, /obj/item/stack))
+			var/datum/component/material_container/mats = GetComponent(/datum/component/material_container)
+			sheets = TRUE
+			for(var/datum/material/mat in D.materials)
+				max_multiplier = min(D.maxstack, round(mats.get_material_amount(mat)/D.materials[mat]))
+			if (max_multiplier>10 && !disabled)
+				m10 = TRUE
+			if (max_multiplier>25 && !disabled)
+				m25 = TRUE
+		else
+			if(!disabled && can_build(D, 5))
+				m5 = TRUE
+			if(!disabled && can_build(D, 10))
+				m10 = TRUE
+			var/datum/component/material_container/mats = GetComponent(/datum/component/material_container)
 
-/obj/machinery/autolathe/Topic(href, href_list)
-	if(..())
+			for(var/datum/material/mat in D.materials)
+				max_multiplier = min(50, round(mats.get_material_amount(mat)/(D.materials[mat] * creation_efficiency)))
+
+		var/list/design = list(
+			name = D.name,
+			id = D.id,
+			ref = REF(src),
+			cost = get_design_cost(D),
+			buildable = unbuildable,
+			mult5 = m5,
+			mult10 = m10,
+			mult25 = m25,
+			mult50 = m50,
+			sheet = sheets,
+			maxmult = max_multiplier,
+		)
+		output += list(design)
+	return output
+
+/obj/machinery/autolathe/ui_act(action, params)
+	. = ..()
+	if(.)
 		return
-	if (!busy)
-		if(href_list["menu"])
-			screen = text2num(href_list["menu"])
-			updateUsrDialog()
-
-		if(href_list["category"])
-			selected_category = href_list["category"]
-			updateUsrDialog()
-
-		if(href_list["make"])
-
+	if(action == "menu")
+		selected_category = null
+		matching_designs.Cut()
+
+	if(action == "category")
+		selected_category = params["selectedCategory"]
+		matching_designs.Cut()
+
+	if(action == "search")
+		matching_designs.Cut()
+
+		for(var/v in stored_research.researched_designs)
+			var/datum/design/D = SSresearch.techweb_design_by_id(v)
+			if(findtext(D.name,params["to_search"]))
+				matching_designs.Add(D)
+	if(action == "make")
+		if (!busy)
 			/////////////////
 			//href protection
-			being_built = stored_research.isDesignResearchedID(href_list["make"])
+			being_built = stored_research.isDesignResearchedID(params["id"])
 			if(!being_built)
 				return
```

Comment:
You must return TRUE at the end of each ui_act branch, otherwise UI will not immediately update.